### PR TITLE
Add ranger recovery logic

### LINKED ISSUE #243: Add header consistency and authority set updates to GRANDPA range relayer
### Item Type

Task

### Description

To ensure the GRANDPA range proofs work, two things are important:
- We need to ensure there is no gap in the header range we want to submit. 
- The authority set must stay updated in the circuit. If the authority set is updated on `target`, we must submit the GRANDPA justification of that block to the circuit, which will update the authority set.

These two things need to be reflected in the grandpa range demo

### Additional Information

_No response_

### Acceptance Criteria

- Authority set changes are handled correctly
- Header gaps are detected and fixed by querying the missing headers. (need to be sequential)

### What are the points assigned to this item?

2

### Is the issue ready?

- [X] This issue is ready

### PR DESCRIPTION
**Pull request by petscheit**
## Summary
The grandpa ranger was missing authority set update detection. This is important in two scenarios:

### While running:
While the ranger is running, it must be ensured justifications containing an authority update are submitted to the circuit. With this PR, every incoming header is checked for an authority set update, and submitted to the circuit if needed.

### On start:
Before starting the listeners, the ranger now checks for any missed justifications containing an authority update. These are then submitted to the circuit, ensuring the following justifications can be submitted.

closes #243 #283

(and I had to change to tab indentions, as this space stuff was making me go insane. Sorry.)

### CODE REVIEW

File: client/grandpa-ranger/src/util.ts [hunk 1]
```diff
@@ -1,85 +1,102 @@
...
 }
 
-export const exec = promisify(_exec)
+// gets events of a specific block, and checks if authoritySetUpdate is included.
+// it would be more efficient to listen for the NewAuthorities event directly, but I'm unable to decode the block number currently.
+export const containsAuthoritySetUpdate= async (api: any, blockNumber: number) => {
+    const hash = await api.rpc.chain.getBlockHash(blockNumber);
+    const notifications = await api.query.system.events.at(hash);
+    return notifications.some(noti => noti.event.method === "NewAuthorities")
+}
```
Comment (petscheit) [inline client/grandpa-ranger/src/util.ts:20]: very inefficient way of checking if a justification contains an authority set update

**Comment by petscheit** [comment]
Sorry about the mess with the indentions. I have commented on the important bits.

File: client/grandpa-ranger/package.json [hunk 2]
```diff
@@ -11,8 +10,8 @@
   "license": "Apache-2.0",
   "dependencies": {
     "@polkadot/api": "=7.14.3",
-    "@t3rn/types": "=0.1.1",
     "@types/debug": "=4.1.7",
+    "axios": "^0.27.2",
```
Comment (chiefbiiko) [inline client/grandpa-ranger/package.json:14]:
```suggestion
    "axios": "=0.27.2",
```
developed dis auto update paranoia..would go for a pinned npm version

**Review by chiefbiiko — DISMISSED** [review]
SUPER SAUCY !!!

File: client/grandpa-ranger/src/index.ts [hunk 3]
```diff
@@ -4,168 +4,183 @@ import RelaychainListener from "./listeners/relaychain"
...
-        let [storageProof, headerNumber] = await relaychain.getStorageProof(
-          data.anchorHash,
-          parachainId
+
+    async initializeEventListeners() {
+        // once a relaychains finality proof has been submitted
+        this.relayer.on("FinalityProofSubmitted", async (data: any) => {
+            InstanceManager.debug("FinalityProofSubmitted")
+
+            const _relaychainParams: {
```
Comment (MaciejBaj) [inline client/grandpa-ranger/src/index.ts:96]: remove underscore since the const is used later

File: client/grandpa-ranger/src/relayer.ts [hunk 4]
```diff
@@ -1,143 +1,179 @@
...
-      anchorNumber: number
-    }[]
-  > {
-    const nonce = await fetchNonce(this.api, keyring.alice.address)
-    Relayer.debug("submitParachainHeaders nonce", nonce.toString())
+        await this.api.tx.multiFinalityVerifierDefault
+            .submitFinalityProof(anchorHeader, justification, gatewayId)
+            .signAndSend(this.signer, { nonce }, async result => {
+                if (result.isError) {
+                    Relayer.debug("FinalityProofSubmitted failed")
```
Comment (MaciejBaj) [inline client/grandpa-ranger/src/relayer.ts:44]: no try error to redo?

File: pallets/circuit-portal/rpc/src/types.rs [hunk 5]
```diff
@@ -65,16 +64,17 @@ pub struct CallRequest<AccountId, Balance> {
...
-pub enum RpcReadLatestGatewayHeight {
-    /// Successful execution
-    Success {
-        /// Output data
-        encoded_height: Bytes,
-    },
-    /// Error execution
-    Error(()),
-}
+// ToDo: the types where not implemented consistently. We need to figure out how to handle RPC in general, currently its handled differently in contracts and RPC.
```
Comment (MaciejBaj) [inline pallets/circuit-portal/rpc/src/types.rs:67]: not a fan of commenting out. just remove if unused

**Review by MaciejBaj — CHANGES_REQUESTED** [review]

**Pull request closed without merging**